doc/networking.md

```diff
@@ -0,0 +1,132 @@
+# Livepeer Networking Protocol v2 Spec
+
+## High level flow
+
+![broadcaster-transcoder network v2 1](https://user-images.githubusercontent.com/292510/41455677-c8437268-7032-11e8-9ce8-bfdd9b6e3fc0.png)
+[Sequence diagram source](https://sequencediagram.org/index.html#initialData=C4S2BsFMAIBkQG6QA6UgJ2gOUsA7gPboDWIAdgObQIBMAUHcgIbqgDGIzZw0ASpBRABnYOgCedAELoCTACZsmIjAFoAfP0EjxALmgB5dGwAWkbU2BE+kAI51Nw0WJXrpshUuAY9hk2dEWVvxCyAxu8orK6Oq+puaW6DoAOmQAKuhMZEJsBHIY1nbhHlEAPC6x-hkJOumZ2bn5AJJkAGYEDAzgBAShRZFe0Wq1WTl5idAAygIAtpDcBVIyEZ4YZSrD9WN6UxSz88Ghc3J0QA)
+
+## Broadcaster to Registry
+
+The broadcaster does an on-chain lookup to retrieve the orchestrator's ServiceURI in order to initiate an transcoding job.
+
+## Broadcaster to Orchestrator:
+
+### gRPC `GetTranscoder`
+
+This is called by the broadcaster to request information on a transcoder from an orchestrator. The request contains the following data:
+
+```protobuf
+// This request is sent by the broadcaster in `GetTranscoder` to request
+// information on which transcoder to use.
+message TranscoderRequest {
+
+  // ID of the job that the broadcaster needs a transcoder for
+  int64 jobId = 1;
+
+  // Broadcaster's signature over the jobId
+  bytes sig = 2;
+}
+```
+
+The `sig` field is the broadcaster's signature using its Ethereum key, over the jobId expressed as a string. The signature consists of the resulting bytes without any further encoding.
+`sig = broadcaster.Sign(jobId.String())`
+
+Verification of `TranscoderRequest` consists of the following steps:
+1. Fetch the job based on JobId.
+    * TODO distinguish if an orchestrator is lagging behind and hasn't yet received the block with the job. Might require incorporating the node's **current** block number into the request. `CreationBlock`  as stored in the job only approximates when the job is submitted to the chain, not when it is included in a block and assigned to a transcoder.
+2. Ensure the orchestrator is assigned to the job.
+3. Ensure the current block falls within the valid range of `CreationBlock` ... `EndBlock` for the job.
+4. Verify that the signature comes from the broadcaster.
+
+The `TranscoderInfo` response contains:
+
+```protobuf
+// The orchestrator sends this in response to `GetTranscoder`, containing the
+// transcoder URI, associated credentials authorizing the broadcaster to
+// use the transcoder, and miscellaneous data related to the job.
+message TranscoderInfo {
+
+  // URI of the transcoder to use for submitting segments
+  string transcoder = 1;
+
+  // Signals the authentication method to expect within `credentials`. This
+  // field is opaque to the broadcaster, and should be passed to the transcoder.
+  string authType = 2;
+
+  // Credentials to verify the request has been authorized by an orchestrator.
+  // This field is opaque to the broadcaster.
+  string credentials = 3;
+
+  // Transcoded streamId list to update the master manifest on the broadcaster.
+  map<string, string> streamIds = 50;
```

Comment:
This is might just be a protocol buffer thing - why is the field number 50 and not 4 (continuing from the previous field number of 3)?

Reply:
In protobufs, field numbers 1-15 take one byte to encode, 16-2047 take two bytes, so the recommendation is to give less frequently occurring fields a higher number, and leave room in the lower end of the range for future expansion. (Why did I pick 50 specifically, then? No real reason other than I couldn't remember what the range was.) https://developers.google.com/protocol-buffers/docs/proto#assigning-field-numbers

In the case of StreamIds -- I wasn't sure if we would always be returning this type of data in future iterations of the protocol, since it's quite specific to how the relay protocol currently works. But perhaps that assumption is a bit premature here, and it's not really implemented consistently for the other structs either. Will see about cleaning it up.